File: src/curator/snapshot.ts
import { pool } from "../db/pool.js";

export type QueueSnapshot = {
  done_unpublished: number;
  done_total: number;
  needs_human_review: number;
  queued: number;
  running: number;
  failed_no_data: number;
  failed_validation: number;
  failed_crashed: number;
};

export async function loadSnapshot(): Promise<QueueSnapshot> {
  const { rows } = await pool.query<{
    done_unpublished: string;
    done_total: string;
    needs_human_review: string;
    queued: string;
    running: string;
    failed_no_data: string;
    failed_validation: string;
    failed_crashed: string;
  }>(`
    SELECT
      COUNT(*) FILTER (WHERE r.status = 'done'
                        AND NOT EXISTS (SELECT 1 FROM publications p WHERE p.run_id = r.id))::text AS done_unpublished,
      COUNT(*) FILTER (WHERE r.status = 'done')::text                AS done_total,
      COUNT(*) FILTER (WHERE r.status = 'needs_human_review')::text  AS needs_human_review,
      COUNT(*) FILTER (WHERE r.status = 'queued')::text              AS queued,
      COUNT(*) FILTER (WHERE r.status = 'running')::text             AS running,
      COUNT(*) FILTER (WHERE r.status = 'failed_no_data')::text      AS failed_no_data,
      COUNT(*) FILTER (WHERE r.status = 'failed_validation')::text   AS failed_validation,
      COUNT(*) FILTER (WHERE r.status = 'failed_crashed')::text      AS failed_crashed
    FROM task_runs r
  `);
  const r = rows[0]!;
  return {
    done_unpublished: parseInt(r.done_unpublished, 10),
    done_total: parseInt(r.done_total, 10),
    needs_human_review: parseInt(r.needs_human_review, 10),
    queued: parseInt(r.queued, 10),
    running: parseInt(r.running, 10),
    failed_no_data: parseInt(r.failed_no_data, 10),
    failed_validation: parseInt(r.failed_validation, 10),
    failed_crashed: parseInt(r.failed_crashed, 10),
  };
}

export function formatSnapshot(s: QueueSnapshot): string {
  return [
    "<queue>",
    `done без публикации: ${s.done_unpublished}`,
    `done всего: ${s.done_total}`,
    `needs_human_review: ${s.needs_human_review}`,
    `queued: ${s.queued}`,
    `running: ${s.running}`,
    `failed_no_data: ${s.failed_no_data}`,
    `failed_validation: ${s.failed_validation}`,
    `failed_crashed: ${s.failed_crashed}`,
    "</queue>",
  ].join("\n");
}

File: src/parts_research/curator/snapshot.py
"""Snapshot очереди для куратора. Подмешивается в начало каждого user-сообщения,
чтобы куратор всегда видел актуальное состояние: счётчики по статусам + последние N
`done`-ранов без публикации (которые обычно и просят «обработать»)."""

from __future__ import annotations

import asyncpg

PENDING_LIMIT = 20


# Считаем по ПОСЛЕДНЕМУ run каждой задачи — консистентно с дашбордом (api/queries.py),
# чтобы число «ждёт куратора» в снапшоте совпадало с чипом в UI (re-run не двоит).
_LATEST = "SELECT DISTINCT ON (task_id) id AS run_id, task_id, status FROM task_runs ORDER BY task_id, id DESC"


async def load_snapshot(pool: asyncpg.Pool) -> dict:
    counts = await pool.fetch(f"SELECT status, count(*) AS n FROM ({_LATEST}) s GROUP BY status")
    pending = await pool.fetch(
        f"SELECT s.run_id, t.article, dp.name, dp.vehicle_classes, dp.is_kit "
        f"FROM ({_LATEST}) s "
        f"JOIN tasks t ON t.id = s.task_id "
        f"LEFT JOIN draft_parts dp ON dp.run_id = s.run_id "
        f"WHERE s.status = 'done' AND NOT EXISTS (SELECT 1 FROM publications p WHERE p.run_id = s.run_id) "
        f"ORDER BY s.run_id LIMIT $1",
        PENDING_LIMIT,
    )
    return {
        "counts": {r["status"]: r["n"] for r in counts},
        "pending": [dict(r) for r in pending],
    }


def format_snapshot(snap: dict) -> str:
    counts = snap["counts"]
    pending = snap["pending"]
    lines = ["<queue>"]
    lines.append("statuses: " + (", ".join(f"{k}={v}" for k, v in sorted(counts.items())) or "—"))
    suffix = f" (показаны первые {PENDING_LIMIT})" if len(pending) == PENDING_LIMIT else ""
    lines.append(f"done без публикации: {len(pending)}{suffix}")
    for p in pending:
        classes = ",".join(p["vehicle_classes"] or []) or "—"
        lines.append(
            f"  run={p['run_id']} {p['article']} | {p['name']} | classes={classes} | kit={p['is_kit']}"
        )
    lines.append("</queue>")
    return "\n".join(lines)
